### PR TITLE
Use ByteStrings more.

### PR DESCRIPTION
**Pull request by kffaxen**

### CODE REVIEW

File: src/Feldspar/Onnx/onnxToFeld.hs [hunk 1]
```diff
@@ -194,39 +199,44 @@ showElemT TD.STRING     = "String"
...
+showDim :: Maybe TSP.Value -> L.ByteString
 showDim (Just (TSP.Dim_value i)) = show i
-showDim (Just (TSP.Dim_param p)) = toStr p
+showDim (Just (TSP.Dim_param p)) = unUtf8 p
 showDim Nothing = "*"
 
-toStr :: H.Utf8 -> String
-toStr (H.Utf8 s) = U.toString s
+-- | Unpack a ByteString from the Utf8 newtype
+unUtf8 :: H.Utf8 -> L.ByteString
```
Comment (pjonsson) [inline src/Feldspar/Onnx/onnxToFeld.hs:220]: Isn't this a cut and paste of `utf8` from Protobufs?

Comment (kffaxen) [reply src/Feldspar/Onnx/onnxToFeld.hs:220]: Of course, but I looked in the wrong place in the documentation and did not find it. Will change.

File: src/Feldspar/Onnx/onnxToFeld.hs [hunk 2]
```diff
@@ -99,49 +104,49 @@ buildValues TD.DOUBLE     t = D.foldMap (\ x -> B.doubleDec x      <> B.string8
...
-            , name ++ " " ++ unwords params ++ " = " ++ tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
+            , name <> " " <> L.unwords params <> " = " <> tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
             , "  where "
-              ++ intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
+              <> L.intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
             ]
   where name = "model" -- strFromJ $ G.name gr
         wName = "weights"
-        initVs = map (toStr . fromJust . TP.name) $ D.toList $ G.initializer gr
+        initVs = map (strFromJ . TP.name) $ D.toList $ G.initializer gr
```
Comment (pjonsson) [inline src/Feldspar/Onnx/onnxToFeld.hs:127]: Can we avoid `strFromJ` here and let `initSet` be the Utf8-strings of names to simplify the code for `inps`?

File: src/Feldspar/Onnx/onnxToFeld.hs [hunk 3]
```diff
@@ -99,49 +104,49 @@ buildValues TD.DOUBLE     t = D.foldMap (\ x -> B.doubleDec x      <> B.string8
...
-         ++ [ "    " ++ mangle (toStr v) | v <- D.toList $ N.input n]
+mkNode :: N.NodeProto -> [L.ByteString]
+mkNode n = [outs <> " = " <> op <> " " <> attrs]
+         ++ [ "    " <> mangle (unUtf8 v) | v <- D.toList $ N.input n]
          ++ [""]
-  where outs = tuplify $ map (mangle . toStr) $ D.toList $ N.output n
-        op = "onnx" ++ toStr (fromJust $ N.op_type n) ++ "_" ++ show (length $ D.toList $ N.input n)
-        attrs = "[" ++ intercalate ", " (map showAttribute $ D.toList $ N.attribute n) ++ "]"
+  where outs = tuplify $ map (mangle . unUtf8) $ D.toList $ N.output n
+        op = "onnx" <> strFromJ (N.op_type n) <> "_" <> show (length $ D.toList $ N.input n)
```
Comment (pjonsson) [inline src/Feldspar/Onnx/onnxToFeld.hs:140]: There is a `D.length` that lets us avoid converting things to a list that we just take the length of.

File: src/Feldspar/Onnx/onnxToFeld.hs [hunk 4]
```diff
@@ -99,49 +104,49 @@ buildValues TD.DOUBLE     t = D.foldMap (\ x -> B.doubleDec x      <> B.string8
...
             , "import Feldspar.Onnx.Operators"
             , ""
-            , "main = program " ++ name
+            , "main = program " <> name
             , ""
-            , name ++ " " ++ unwords params ++ " = " ++ tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
+            , name <> " " <> L.unwords params <> " = " <> tuplify (map (mangle . vipName) $ D.toList $ G.output gr)
             , "  where "
-              ++ intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
+              <> L.intercalate "\n        " ("-- Nodes " : concatMap mkNode nodes)
```
Comment (pjonsson) [inline src/Feldspar/Onnx/onnxToFeld.hs:123]: Can we avoid the list conversion by using `D.concatMap`?

File: src/Feldspar/Onnx/onnxToFeld.hs (hunk 3 above)
Comment (kffaxen) [reply src/Feldspar/Onnx/onnxToFeld.hs:140]: Ok!

File: src/Feldspar/Onnx/onnxToFeld.hs (hunk 4 above)
Comment (kffaxen) [reply src/Feldspar/Onnx/onnxToFeld.hs:123]: I'm not sure I follow you here. L.intercalate really wants a list of ByteStrings.

Comment (pjonsson) [reply src/Feldspar/Onnx/onnxToFeld.hs:122]: Replace `concatMap` with `D.concatMap` and remove `D.toList` from the definition of `nodes`?

Comment (kffaxen) [reply src/Feldspar/Onnx/onnxToFeld.hs:122]: Sure. I thought Data.Foldable.concatMap would use the foldable type for the resulting list, not just for the input list.

**Pull request merged**